# Setup

## Platform

|setting  |value                        |
|:--------|:----------------------------|
|version  |R version 3.4.1 (2017-06-30) |
|system   |x86_64, linux-gnu            |
|ui       |X11                          |
|language |en                           |
|collate  |en_US.UTF-8                  |
|tz       |America/Los_Angeles          |
|date     |2017-09-12                   |

## Packages

|package        |*  |version    |date       |source                                |
|:--------------|:--|:----------|:----------|:-------------------------------------|
|aroma.core     |   |3.1.0-9000 |2017-09-12 |local (HenrikBengtsson/aroma.core@NA) |
|aroma.light    |   |3.6.0      |2017-09-12 |cran (@3.6.0)                         |
|Cairo          |   |1.5-9      |2015-09-26 |cran (@1.5-9)                         |
|DNAcopy        |   |1.50.1     |2017-09-12 |cran (@1.50.1)                        |
|EBImage        |   |4.18.2     |2017-09-12 |cran (@4.18.2)                        |
|expectile      |   |0.3.0      |2017-09-12 |cran (@0.3.0)                         |
|future         |   |1.6.1      |2017-09-09 |cran (@1.6.1)                         |
|GLAD           |   |2.40.0     |2017-09-12 |cran (@2.40.0)                        |
|HaarSeg        |   |0.0.3      |2017-09-12 |cran (@0.0.3)                         |
|listenv        |   |0.6.0      |2015-12-28 |cran (@0.6.0)                         |
|matrixStats    |   |0.52.2     |2017-04-14 |cran (@0.52.2)                        |
|mpcbs          |   |1.1.1      |2017-09-12 |cran (@1.1.1)                         |
|png            |   |0.1-8      |2017-09-12 |cran (@0.1-8)                         |
|preprocessCore |   |1.38.1     |2017-09-12 |cran (@1.38.1)                        |
|PSCBS          |   |0.63.0     |2017-06-28 |cran (@0.63.0)                        |
|R.cache        |   |0.12.0     |2015-11-12 |cran (@0.12.0)                        |
|R.devices      |   |2.15.1     |2016-11-10 |cran (@2.15.1)                        |
|R.filesets     |   |2.11.0     |2017-02-28 |cran (@2.11.0)                        |
|R.methodsS3    |   |1.7.1      |2016-02-16 |cran (@1.7.1)                         |
|R.oo           |   |1.21.0     |2016-11-01 |cran (@1.21.0)                        |
|R.rsp          |   |0.41.0     |2017-04-16 |cran (@0.41.0)                        |
|R.utils        |   |2.5.0      |2016-11-07 |cran (@2.5.0)                         |
|RColorBrewer   |   |1.1-2      |2014-12-07 |cran (@1.1-2)                         |
|sfit           |   |0.3.1      |2017-09-12 |cran (@0.3.1)                         |

# Check results

7 packages with problems

|package          |version | errors| warnings| notes|
|:----------------|:-------|------:|--------:|-----:|
|ACNE             |0.8.1   |      0|        1|     0|
|aroma.affymetrix |3.1.0   |      0|        1|     0|
|aroma.cn         |1.6.1   |      0|        1|     0|
|calmate          |0.12.1  |      0|        1|     0|
|MPAgenomics      |1.1.2   |      0|        1|     2|
|NSA              |0.0.32  |      0|        1|     6|
|Repitools        |1.23.0  |      1|        0|     1|

## ACNE (0.8.1)
Maintainer: Henrik Bengtsson <henrikb@braju.com>  
Bug reports: https://github.com/HenrikBengtsson/ACNE/issues

0 errors | 1 warning  | 0 notes

```
checking CRAN incoming feasibility ... WARNING
Maintainer: ‘Henrik Bengtsson <henrikb@braju.com>’

Insufficient package version (submitted: 0.8.1, existing: 0.8.1)

Possibly mis-spelled words in DESCRIPTION:
  NMF (16:157)
  Summarization (9:29)
  summarization (16:16)

The Date field is over a month old.

This build time stamp is over a month old.
```

## aroma.affymetrix (3.1.0)
Maintainer: Henrik Bengtsson <henrikb@braju.com>  
Bug reports: https://github.com/HenrikBengtsson/aroma.affymetrix/issues

0 errors | 1 warning  | 0 notes

```
checking CRAN incoming feasibility ... WARNING
Maintainer: ‘Henrik Bengtsson <henrikb@braju.com>’

Insufficient package version (submitted: 3.1.0, existing: 3.1.0)

The Date field is over a month old.

This build time stamp is over a month old.
```

## aroma.cn (1.6.1)
Maintainer: Henrik Bengtsson <henrikb@braju.com>  
Bug reports: https://github.com/HenrikBengtsson/aroma.cn/issues

0 errors | 1 warning  | 0 notes

```
checking CRAN incoming feasibility ... WARNING
Maintainer: ‘Henrik Bengtsson <henrikb@braju.com>’

Insufficient package version (submitted: 1.6.1, existing: 1.6.1)

Possibly mis-spelled words in DESCRIPTION:
  MSCN (16:71)
  TumorBoost (18:41)

The Date field is over a month old.

This build time stamp is over a month old.
```

## calmate (0.12.1)
Maintainer: Henrik Bengtsson <henrikb@braju.com>  
Bug reports: https://github.com/HenrikBengtsson/calmate/issues

0 errors | 1 warning  | 0 notes

```
checking CRAN incoming feasibility ... WARNING
Maintainer: ‘Henrik Bengtsson <henrikb@braju.com>’

Insufficient package version (submitted: 0.12.1, existing: 0.12.1)

Possibly mis-spelled words in DESCRIPTION:
  ASCNs (10:93)

The Date field is over a month old.

This build time stamp is over a month old.
```

## MPAgenomics (1.1.2)
Maintainer: Samuel Blanck <samuel.blanck@inria.fr>

0 errors | 1 warning  | 2 notes

```
checking CRAN incoming feasibility ... WARNING
Maintainer: ‘Samuel Blanck <samuel.blanck@inria.fr>’

Insufficient package version (submitted: 1.1.2, existing: 1.1.2)

Possibly mis-spelled words in DESCRIPTION:
  MPAgenomics (10:55, 15:23)
  Preprocess (10:14)

No package encoding and non-ASCII characters in the following R files:
  R/bivariateSignal.R
    64: # etrange=c(2,36)#66 semble greff<c3><a9> #7 CN <c3><a9>trange
  R/cghcall_private.R
    140:     #TODO v<c3><a9>rifier pertinence correction bug avec segment extreme d'un seul point

The Date field is over a month old.

This build time stamp is over a month old.

checking dependencies in R code ... NOTE
'library' or 'require' calls in package code:
  ‘R.devices’ ‘R.filesets’ ‘R.methodsS3’ ‘R.oo’ ‘aroma.affymetrix’
  ‘aroma.cn’ ‘aroma.core’ ‘aroma.light’ ‘matrixStats’ ‘snowfall’
  Please use :: or requireNamespace() instead.
  See section 'Suggested packages' in the 'Writing R Extensions' manual.
Unexported object imported by a ':::' call: ‘cghseg:::segmeanCO’
  See the note in ?`:::` about the use of this operator.

checking R code for possible problems ... NOTE
.varregtimescount: no visible global function definition for ‘var’
CGHSEGaroma: no visible global function definition for ‘read.csv’
CGHSEGaroma : <anonymous>: no visible global function definition for
  ‘points’
CGHSEGaroma : <anonymous>: no visible global function definition for
  ‘lines’
CGHSEGaroma : <anonymous>: no visible global function definition for
  ‘write.table’
CGHcall: no visible global function definition for ‘mad’
... 43 lines ...
tumorboostPlot: no visible global function definition for ‘par’
tumorboostPlot: no visible global function definition for ‘axis’
tumorboostPlot: no visible global function definition for ‘points’
Undefined global functions or variables:
  axis head lines lm mad median optim par points read.csv sd var
  write.table
Consider adding
  importFrom("graphics", "axis", "lines", "par", "points")
  importFrom("stats", "lm", "mad", "median", "optim", "sd", "var")
  importFrom("utils", "head", "read.csv", "write.table")
to your NAMESPACE file.
```

## NSA (0.0.32)
Maintainer: Maria Ortiz-Estevez <mortizest@gmail.com>

0 errors | 1 warning  | 6 notes

```
checking CRAN incoming feasibility ... WARNING
Maintainer: ‘Maria Ortiz-Estevez <mortizest@gmail.com>’

Insufficient package version (submitted: 0.0.32, existing: 0.0.32)

The Title field should be in title case, current version then in title case:
‘Post-normalization of total copy numbers’
‘Post-Normalization of Total Copy Numbers’

The Date field is over a month old.

This build time stamp is over a month old.

checking package dependencies ... NOTE
Depends: includes the non-default packages:
  ‘R.methodsS3’ ‘MASS’ ‘matrixStats’ ‘R.oo’ ‘R.utils’ ‘aroma.core’
  ‘aroma.affymetrix’ ‘DNAcopy’
Adding so many packages to the search path is excessive and importing
selectively is preferable.

checking top-level files ... NOTE
Non-standard file/directory found at top level:
  ‘incl’

checking dependencies in R code ... NOTE
Packages in Depends field not imported from:
  ‘DNAcopy’ ‘MASS’ ‘R.methodsS3’ ‘R.oo’ ‘aroma.affymetrix’ ‘aroma.core’
  ‘matrixStats’
  These packages need to be imported from (in the NAMESPACE file)
  for when this namespace is loaded but not attached.

checking S3 generic/method consistency ... NOTE
Found the following apparent S3 methods exported but not registered:
  NSAByTotalAndFracB.matrix allocateOutputDataSets.NSANormalization
  allocateOutputDataSets.SNPsNormalization
  allocateOutputDataSets.SampleNormalization
  as.character.NSANormalization as.character.SNPsNormalization
  as.character.SampleNormalization findArraysTodo.NSANormalization
  findArraysTodo.SampleNormalization findUnitsTodo.SNPsNormalization
  fitNSA.matrix fitNSAcnPs.matrix getDataSets.NSANormalization
  getDataSets.SNPsNormalization getDataSets.SampleNormalization
  getName.NSANormalization getName.SNPsNormalization
  getName.SampleNormalization getOutputDataSets.NSANormalization
  getOutputDataSets.SNPsNormalization
  getOutputDataSets.SampleNormalization getPath.NSANormalization
  getPath.SNPsNormalization getPath.SampleNormalization
  getRootPath.NSANormalization getRootPath.SNPsNormalization
  getRootPath.SampleNormalization process.NSANormalization
  process.SNPsNormalization process.SampleNormalization
  sampleNByTotalAndFracB.numeric snpsNByTotalAndFracB.matrix
See section ‘Registering S3 methods’ in the ‘Writing R Extensions’
manual.

checking R code for possible problems ... NOTE
NB: .First.lib is obsolete and will not be used in R >= 3.0.0

.First.lib: no visible global function definition for
  ‘packageDescription’
NSAByTotalAndFracB.matrix: no visible global function definition for
  ‘throw’
NSAByTotalAndFracB.matrix: no visible global function definition for
  ‘str’
NSANormalization: no visible global function definition for ‘throw’
... 279 lines ...
  extractMatrix findUnitsTodo getAsteriskTags getChipType getFile
  getFullName getFullNames getGenomeInformation getName getNames
  getPath getPathname getPathnames getPositions getRam getRootPath
  getTags getUnitsOnChromosome hist median nbrOfFiles newInstance
  packageDescription rowAlls rowMedians segment setTags str throw trim
  verbose
Consider adding
  importFrom("graphics", "hist")
  importFrom("stats", "approxfun", "median")
  importFrom("utils", "packageDescription", "str")
to your NAMESPACE file.

checking Rd line widths ... NOTE
Rd file 'NSANormalization.Rd':
  \examples lines wider than 100 characters:
     by <- 50e3; # 50kb bins; you may want to try with other amounts of smoothing xOut <- seq(from=xRange[1], to=xRange[2], by=by);
     plot(getSignals(cnCNPS), getSignals(cnSNPS), xlim=Clim, ylim=Clim); abline(a=0, b=1, col="red", lwd=2);

These lines will be truncated in the PDF manual.
```

## Repitools (1.23.0)
Maintainer: Mark Robinson <mark.robinson@imls.uzh.ch>

1 error  | 0 warnings | 1 note 

```
checking whether package ‘Repitools’ can be installed ... ERROR
Installation failed.
See ‘/home/hb/repositories/aroma.core/revdep/checks/Repitools.Rcheck/00install.out’ for details.

checking CRAN incoming feasibility ... NOTE
Maintainer: ‘Mark Robinson <mark.robinson@imls.uzh.ch>’

Package duplicated from https://bioconductor.org/packages/3.5/bioc

Possibly mis-spelled words in DESCRIPTION:
  BayMeth (22:9)
  Epigenomic (4:8)
  epigenomic (18:57, 20:9)
  methylation (21:50, 22:33)
  summarization (19:33)

Found the following (possibly) invalid URLs:
  URL: http://www.bioinformatics.bbsrc.ac.uk/projects/fastqc/
    From: man/genQC.Rd
    Status: 403
    Message: Forbidden

The Title field should be in title case, current version then in title case:
‘Epigenomic tools’
‘Epigenomic Tools’

The Date field is not in ISO 8601 yyyy-mm-dd format.

This build time stamp is over a month old.
```

